Automattic/jetpack
# Jetpack Search/Sync: ensure syncing of post meta and taxonomies that are indexed

Our search index includes a fixed list of post meta keys and custom taxonomies in order to limit the number of fields in the index (see documentation https://jetpack.com/support/search/frequently-asked-questions/#troubleshoot-custom-tax).

We have so far not been syncing the list from wp.com into Jetpack which has had many downsides:
- Uncertainty if it worked in all cases (also lack of tests)
- No easy way to add new fields
- No visibility into what is indexed
- No way to double check that the config that the user has will work.
- No way to provide post meta faceting in search

This fixes all of that.

The class-search.php file will be replacing the old-style class.jetpack-sync-module-search.php that we are currently using on wp.com for the indexing. See https://github.com/Automattic/wpes-lib/blob/master/src/indices/jetpack-search-v2-index.php#L842

Fixes #15672
Fixes #8663

#### Does this pull request change what data or activity we track or use?

This potentially expands the list of post meta that we sync a lot. I don't think that is a real change to what we need to document though.

#### Testing instructions:

I tested this with a new set of unit tests. The unit tests are one of the more important pieces since I expect that the list of post meta and taxonomies will evolve over time and we want to make sure that it is trivial to add a new field.

Warning: this adds 1608 tests to the sync testsuite. They take almost 5 minutes to run. I've already spent a bit of time trying to reduce the time though (I cut out doing any update/delete tests). I think checking every item is worth it unfortunately. I found a number of keys that just don't work or that are weird (hello post_format taxonomy!).

```
yarn docker:phpunit --filter=WP_Test_Jetpack_Sync_Search
yarn docker:phpunit --testsuite=sync
```

#### Proposed changelog entry for your changes:
* Search: adds explicit post meta keys and taxonomies to sync into the index. Particularly important for integrating with third-party plugins.
